control-plane/agents/src/bin/core/volume/operations.rs

```diff
@@ -980,49 +888,60 @@ impl CreateVolumeExe for CreateVolume {
+                if replicas.len() >= self.replicas as usize {
+                    break;
+                } else if replicas.iter().any(|r| {
+                    r.node == replica.node
+                        || spread_label_is_same(r, replica, context).unwrap_or_else(|error| {
+                            context.volume.error(&format!(
+                                "Failed to create replica {:?} for volume, error: {}",
+                                replica,
+                                error.full_string()
+                            ));
+                            false
+                        })
+                }) {
+                    // don't re-use the same node or same exclusion labels
+                    continue;
+                }
+                let replica = if replicas.is_empty() {
+                    let mut replica = replica.clone();
+                    // the local replica needs to be connected via "bdev:///"
+                    replica.share = Protocol::None;
+                    replica
+                } else {
+                    replica.clone()
+                };
+                match OperationGuardArc::<ReplicaSpec>::create(context.registry, &replica).await {
+                    Ok(replica) => {
+                        replicas.push(replica);
+                    }
+                    Err(error) => {
                         context.volume.error(&format!(
                             "Failed to create replica {:?} for volume, error: {}",
                             replica,
                             error.full_string()
                         ));
-                        false
-                    })
-            }) {
-                // don't re-use the same node or same exclusion labels
-                continue;
+                        // continue trying...
+                    }
+                };
             }
-            let replica = if replicas.is_empty() {
-                let mut replica = replica.clone();
-                // the local replica needs to be connected via "bdev:///"
-                replica.share = Protocol::None;
-                replica
+
+            if replicas.len() < context.volume.as_ref().num_replicas as usize {
+                self.undo(context, replicas).await;
+                continue;
             } else {
-                replica.clone()
-            };
-            match OperationGuardArc::<ReplicaSpec>::create(context.registry, &replica).await {
-                Ok(replica) => {
-                    replicas.push(replica);
-                }
-                Err(error) => {
-                    context.volume.error(&format!(
-                        "Failed to create replica {:?} for volume, error: {}",
-                        replica,
-                        error.full_string()
-                    ));
-                    // continue trying...
-                }
-            };
+                fina_replicas = replicas;
```

Comment:
Why do we need this final_replicas?

Reply:
This is the final set of replicas on which volumes are provisioned.